### PR TITLE
add elixir phoenix example

### PR DESCRIPTION
**Pull request by joecohens**
https://elixir-phoenix-nvsrmcscbb.now.sh

### CODE REVIEW

**Comment by TooTallNate** [comment]
screen shot 2018-08-21 at 12 41 34 pm

Seems like the build is failing on this one 🤔

**Comment by jlirochon** [comment]
Hey @joecohens !

I have a slightly more complicated example where I use [distillery](https://github.com/bitwalker/distillery) to create a release. This way the image is `51.4MB` instead of `114MB`.

Not sure if `now-examples` should be as simple as possible, or do their best to minimize image size. Anyway here is the code : https://github.com/jlirochon/elixir-phoenix-now

It's running on https://now-kfkooukztl.now.sh/

**Comment by TooTallNate** [comment]
> Not sure if `now-examples` should be as simple as possible, or do their best to minimize image size. 

The emphasis is indeed on minimizing image size. Smaller images translates to faster cold boot times, so that's what we are trying to demonstrate here.

**Comment by joecohens** [comment]
Thanks @jlirochon ! I'll look into it :). 

@TooTallNate I agree, my only concern it's people that use frameworks are used to their directory, literally and maybe unfortunately people know they are working on a full featured framework by just looking at the directory structure and base files. But if you agree I can strip everything that is not important for `now-examples` to just spin up one route without all the noise of the boilerplate that frameworks provide. This will be also true for the RoR and Laravel examples.

**Comment by jlirochon** [comment]
@TooTallNate looks like in this case boot time is slower despite the image is smaller, can you confirm ? Seems that boot time of the app > boot time of the container.

@joecohens I think your example is good because it's the minimal hello page that you get when you follow phoenix doc. It makes no sense to disable all the stuff just to prove better performance. What should be done instead is removing things not required to run the app, like build tools for example.

File: elixir-phoenix/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+ADD . .
+
+RUN mix do compile, phx.digest
+
+USER default
+
+EXPOSE 8080
+
+CMD ["mix", "phx.server"]
```
Comment (TooTallNate) [inline elixir-phoenix/Dockerfile]: Is there anything that can be cleaned up / removed at this point before the image in finalized?

Comment (jlirochon) [reply elixir-phoenix/Dockerfile]: The base image, [bitwalker/alpine-elixir:1.7.2](https://hub.docker.com/r/bitwalker/alpine-elixir/) is `105MB` uncompressed by itself. The final image is `114MB` uncompressed. We could remove the source code of the app, but it won't change much. Maybe we can use the official [elixir:1.7.2-alpine](https://hub.docker.com/_/elixir/) which is `80.8MB`.

I achieved `54.4MB` on my final image by starting from `alpine:3.8` and using [bitwalker/distillery](https://github.com/bitwalker/distillery) to create an OTP release. The OTP release includes compiled binaries of the example app, along with the required Erlang/Elixir runtime (not the full distribution).

Comment (joecohens) [reply elixir-phoenix/Dockerfile]: I'm working on the new config right now, I'll let you know how it goes. Thanks!

**Comment by joecohens** [comment]
@TooTallNate @jlirochon Ok, I've updated the build process to use destillery which it happens to have docs for it lol https://hexdocs.pm/distillery/guides/phoenix_walkthrough.html, https://hexdocs.pm/distillery/guides/working_with_docker.html. Also our image went down from 61.4M to 47.5M, the only thing missing it's that for some reason `now` it's not picking up the `CMD`, I'll investigate. 

PS: I've already tried with @jlirochon hack for the port.

**Comment by jlirochon** [comment]
@joecohens I see no problems with `CMD` in my logs when I deploy your code.

The 2 problems I can see are:
- missing `prod.secret.exs` (maybe comment out the line in `prod.exs` ?)
- wrong port detected by the automatic port discovery

Can you point me to the problem you see with `CMD` ?

**Comment by joecohens** [comment]
@jlirochon yeah it wasn't a problem with the `CMD` it's a thing with port discovery and `epmd`

> missing prod.secret.exs (maybe comment out the line in prod.exs ?)
Yep I've commented this locally 🙈 

> wrong port detected by the automatic port discovery
`now` is detecting `epmd` we can try to disable it but it seems a bit hacky lol, I was trying to find other solutions.

Let me know if you have more findings on this.

Thanks!!!

**Comment by joecohens** [comment]
@jlirochon I agree with you, we can add this workaround for now. Thank you very much for your complete reply and help! I've updated the example disabling epmd followed by your example 😄 

https://elixir-phoenix-bmsvydokqe.now.sh

@TooTallNate Let us know what you think and if this looks ok we can proceed.

**Comment by TooTallNate** [comment]
Very nice! Thank you @joecohens, and thanks for the input @jlirochon.

**Pull request merged**